GraphQL Query Analysis feature Documentation (#1443)
wso2/docs-apim: en/docs/learn/api-security/graphql-query-analysis/query-complexity-limitation.md

Often, limiting the depth of a query only is not sufficient to protect a GraphQL service from complex queries. 
This is due to some fields in a GraphQL schema being more costly to compute than others. WSO2 API-Manager introduces 
**“Query Complexity Limitation”** to address such cases. 

With this strategy, a request allowed or rejected based on the complexity of the query, and the configured max complexity 
value of the subscription policy for the corresponding API.

Here we introduced the **complexity values for each of the Fields** in the schema. That **describes the 
computation cost of resolving the particular field**. 

<html>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>
If the field needs to call an expensive service to resolve itself, 
then the **complexity should be at a high level** but if the field is easy to resolve and not an expensive operation, 
the **complexity should be at a low level**. If **no complexity defined** for a field, **by default it will get a value of 1.** 
</p>
</div> 
</html>

> In run time, we calculate the requested query complexity value. When it exceeds the max complexity, 

Reviewer: ```suggestion
In run time, requested query complexity value will be calculated and blocked if it exceeds the max complexity, 
```

then the query will be **rejected.**

> Here is a simple example in which you can understand how to calculate query complexity values before request the query.

Reviewer: ```suggestion
Here is a simple example in which can be used to understand how the calculation happens.
```

> If there is, no arguments provides in your requested query, can calculate complexity value simply 

Reviewer: ```suggestion
If there are, no arguments provides in your requested query, can calculate complexity value simply by
```

adding all the complexity values.

   ```
      query {
       	   allFilms{                   # complexity 1
         		id                     # complexity 3
        		title                  # complexity 1
         		planets {              # complexity 2
           			climate            # complexity 1
         		}
       	   }
      }
      
      # total complexity = 1 + 2 + 1 + 3 + 1 = 8

   ```

> With arguments, you can calculate complexity values as follows.

Reviewer: ```suggestion
With arguments, complexity can be calculated values as follows.
```
